### PR TITLE
Update Elasticsearch to 7.9.1

### PR DESCRIPTION
**Pull request by PrajaktaPurohit**
https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/2017

NOTE: eabee67f3b2ebeaee00cbeb07e073633660c4a88 currently removes compatibility with the older versions.
We need to support older versions as well.

### CODE REVIEW

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb
Comment (PrajaktaPurohit) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb:118]: it might be ok to leave the object index here.

File: src/oc_erchef/apps/chef_index/src/chef_index_expand.erl
Comment (PrajaktaPurohit) [inline src/oc_erchef/apps/chef_index/src/chef_index_expand.erl:140]: L265 of this file uses path as /chef/object. That might need to be replaced with `ok = chef_index_http:delete("/chef/_doc/" ++ Id, [], ?JSON_HEADER),`

**Comment by jashaik** [comment]
Adding the pipeline status - 
https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/2048#83290ef9-0f2c-45fe-b076-7dc0103a4acd

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb
Comment (tas50) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb:84]:
```suggestion
      raise "Unsupported elasticsearch version of #{version}. There is currently support for the major versions of 2, 5 and 6." unless [2, 5, 6, 7].include?(version)
```

Comment (PrajaktaPurohit) [reply omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb]: s/There is currently support for the major versions of 2, 5 and 6./There is currently support for the major versions of 2, 5 6 and 7.

File: oc-chef-pedant/lib/pedant/rspec/search_util.rb
Comment (PrajaktaPurohit) [inline oc-chef-pedant/lib/pedant/rspec/search_util.rb:722]: Is this trying to fix for the case when there is data in the system?

Comment (jashaik) [reply oc-chef-pedant/lib/pedant/rspec/search_util.rb:722]: get_response_count function is to return the number documents found.There is a response structure change for  total field in the ES7.
`"hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "chef",
        "_type" : "_doc",
        "_id" : "2bed8b78e2e51e0c13c82bbbf07f0ad1",
        "_score" : 1.0
      }]}`

in previous versions, Total filed is directly a number indicating the number of documents found.So Added another simple condition to check whether it is a another json or not.

File: src/oc_erchef/apps/chef_index/test/chef_index_expand_tests.erl
Comment (PrajaktaPurohit) [inline src/oc_erchef/apps/chef_index/test/chef_index_expand_tests.erl:329]: nice!

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb
Comment (PrajaktaPurohit) [inline omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb:26]: on the similar lines there are changes in jvm opts (elasticsearch_jvm.opts.erb) fir elasticsearch 7. I wonder if it might be useful to add a separate templates for es7 and have the recipe pick the correct template based on the version of es.

Comment (PrajaktaPurohit) [reply omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb:26]: It might be helpful to include context for this in comments. These setting are applicable only for the users using standalone or tiered configurations. For BYOD(external elasticsearch) none of the omnibus configs are applicable.

File: omnibus_overrides.rb
Comment (PrajaktaPurohit) [inline omnibus_overrides.rb:13]: to be removed before merging.

File: omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb
Comment (PrajaktaPurohit) [inline omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb:176]: Is this required?

**Comment by PrajaktaPurohit** [comment]
Have you had a chance to run the umbrella pipeline by setting ELASTIC_VERSION=7? The default setting is to run with Elasticsearch 6 for external elasticsearch.

**Comment by PrajaktaPurohit** [comment]
The log files at `/var/log/opscode/elasticsearch/current` also lists the deprecation warnings. This might be a good time to  get the fixes in (if any)

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb
Comment (PrajaktaPurohit) [reply omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb:26]: Looks like chef-server lets the user set the listen address for elasticsearch: https://github.com/chef/chef-server/blob/master/omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb#L170 which is used in the elasticsearch.yml. So in theory this could be a multicluster setup - although I do not think that is something we advertise / support at this point.  It might be worth making a note in the docs that cluster for es will not be supported, if they want to use a cluster they can setup external es.

File: omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb
Comment (jashaik) [reply omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb:176]: yes, For identifing the ES version before installation used in elasticsearch.yml.erb file.

**Comment by jashaik** [comment]
> Have you had a chance to run the umbrella pipeline by setting ELASTIC_VERSION=7? The default setting is to run with Elasticsearch 6 for external elasticsearch.

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb
Comment (jashaik) [reply omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb:26]: Sure, I will add comments and update docs.

File: omnibus_overrides.rb
Comment (jashaik) [reply omnibus_overrides.rb:13]: Yes, I have it in my TODO list

File: oc-chef-pedant/lib/pedant/rspec/search_util.rb
Comment (lbakerchef) [reply oc-chef-pedant/lib/pedant/rspec/search_util.rb:722]: With the `and`, I think parentheses aren't necessary.  But if you had used `&&`, they would be necessary:
```
irb(main):061:0> true and {}.is_a? Hash
=> true
irb(main):062:0> false and {}.is_a? Hash
=> false
irb(main):063:0> false or {}.is_a? Hash
=> true
irb(main):064:0> true && {}.is_a? Hash
Traceback (most recent call last):
        1: from C:/Ruby25-x64/bin/irb.cmd:19:in `<main>'
SyntaxError ((irb):64: syntax error, unexpected tCONSTANT, expecting end-of-input)
true && {}.is_a? Hash
                 ^~~~
irb(main):065:0> true && ({}.is_a? Hash)
=> true
```
Up to you if you consider parentheses more readable.

**Review by lbakerchef — COMMENTED** [review]
Looks ok to me.

File: omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb
Comment (PrajaktaPurohit) [reply omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb:176]: Here is what I was thinking about: https://github.com/chef/chef-server/blob/master/omnibus/files/private-chef-cookbooks/private-chef/libraries/helper.rb#L58 detects the version of elasticsearch (after it is running) and then created the index accordingly. But in case of es7, we seem to require this information inorder to start es.

The auto-detection is helpful to make sure anyone running external elasticsearch does not run an unsupported version of elasticsearch. Whether is should be in that helper library or a preflight check is perhaps a question for a different time.

But since we are already doing that - I think of this parameter as redundant. Would es 2,5 and 6 continue to work if the new settings for 7 were made default or does that block from them starting up?

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb
Comment (PrajaktaPurohit) [reply omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb:27]: Are there any change to elasticsearch_jvm.opts?

**Comment by jashaik** [comment]
> Have you had a chance to run the umbrella pipeline by setting ELASTIC_VERSION=7? The default setting is to run with Elasticsearch 6 for external elasticsearch.

Here is the link for external elasticserach version 7 - https://buildkite.com/chef/chef-umbrella-master-chef-server/builds/276#787cb2a9-bfbd-4b1c-a4bb-ea6a63c62689

**Comment by jashaik** [comment]
New pipeline reports after adding ES7 JVM options
Adhoc - https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/2214

File: omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb
Comment (PrajaktaPurohit) [reply omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb:176]: We should add this to the docs.

Comment (jashaik) [reply omnibus/files/private-chef-cookbooks/private-chef/attributes/default.rb:176]: Updated the docs

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb
Comment (PrajaktaPurohit) [inline omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch.yml.erb:26]: It might be useful to leave a note explaining that we have not enabled bootstrap checks, so failing bootstrap checks will appear as warnings in the es logs. https://www.elastic.co/guide/en/elasticsearch/reference/current/bootstrap-checks.html

File: docs-chef-io/content/server/v14/config_rb_server_optional_settings.md
Comment (PrajaktaPurohit) [inline docs-chef-io/content/server/v14/config_rb_server_optional_settings.md:1511]: This will only be supported until we finish upgrading internal es to 7, correct?

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch_jvm_es7.opts.erb
Comment (marcparadise) [inline omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch_jvm_es7.opts.erb:1]: I have not verified these, but are they based on prior version opts? And have we looked to see if we need to change anything based on what's new in ES7?

**Review by marcparadise — APPROVED** [review]
I'm surprised we were still doing 'norms' => { enabled: false } in es6 without problems, it looks like the syntax has been 'norms: false' since v5.0.

These changes look good. I left one question inline; assuming there is no further work to do there, I'm +1 to merge.

**Comment by jashaik** [comment]
New pipelines after commenting the es_version in default.rb file
Umbrella - https://buildkite.com/chef/chef-umbrella-master-chef-server/builds/294
Adhoc - https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/2251

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch_jvm_es7.opts.erb
Comment (jashaik) [reply omnibus/files/private-chef-cookbooks/private-chef/templates/default/elasticsearch_jvm_es7.opts.erb:1]: Yes, We are referring the latest ES7 JVM configurations - https://github.com/elastic/elasticsearch/blob/7.9/distribution/src/config/jvm.options

**Review by PrajaktaPurohit — COMMENTED** [review]
Cannot approve this PR. It looks good to me! Thanks!

**Pull request merged**
